core/src/main/java/org/bitcoinj/store/SPVBlockStore.java

```diff
@@ -175,89 +188,50 @@ private void initNewStore(NetworkParameters params) throws Exception {
-    }
-
     @Override
     public void put(StoredBlock block) throws BlockStoreException {
         final MappedByteBuffer buffer = this.buffer;
         if (buffer == null) throw new BlockStoreException("Store closed");
-
         lock.lock();
         try {
             int cursor = getRingCursor(buffer);
-            if (cursor == fileLength) {
-                // Wrapped around.
-                cursor = FILE_PROLOGUE_BYTES;
-            }
             buffer.position(cursor);
             Sha256Hash hash = block.getHeader().getHash();
-            notFoundCache.remove(hash);
+            chainMaps.put(hash, new ChainMap(cursor, cursor + 128));
             buffer.put(hash.getBytes());
             block.serializeCompact(buffer);
+            height++;
             setRingCursor(buffer, buffer.position());
-            blockCache.put(hash, block);
-        } finally { lock.unlock(); }
+        } finally {
+            lock.unlock();
+        }
     }
 
     @Override
     @Nullable
     public StoredBlock get(Sha256Hash hash) throws BlockStoreException {
+        lock.lock();
         final MappedByteBuffer buffer = this.buffer;
         if (buffer == null) throw new BlockStoreException("Store closed");
-
-        lock.lock();
-        try {
-            StoredBlock cacheHit = blockCache.get(hash);
-            if (cacheHit != null)
-                return cacheHit;
-            if (notFoundCache.get(hash) != null)
-                return null;
-
-            // Starting from the current tip of the ring work backwards until we have either found the block or
-            // wrapped around.
-            int cursor = getRingCursor(buffer);
-            final int startingPoint = cursor;
-            final byte[] targetHashBytes = hash.getBytes();
-            byte[] scratch = new byte[32];
-            do {
-                cursor -= RECORD_SIZE;
-                if (cursor < FILE_PROLOGUE_BYTES) {
-                    // We hit the start, so wrap around.
-                    cursor = fileLength - RECORD_SIZE;
-                }
-                // Cursor is now at the start of the next record to check, so read the hash and compare it.
-                buffer.position(cursor);
-                buffer.get(scratch);
-                if (Arrays.equals(scratch, targetHashBytes)) {
-                    // Found the target.
-                    StoredBlock storedBlock = StoredBlock.deserializeCompact(params, buffer);
-                    blockCache.put(hash, storedBlock);
-                    return storedBlock;
-                }
-            } while (cursor != startingPoint);
-            // Not found.
-            notFoundCache.put(hash, NOT_FOUND_MARKER);
-            return null;
-        } catch (ProtocolException e) {
-            throw new RuntimeException(e);  // Cannot happen.
-        } finally { lock.unlock(); }
+        ChainMap map = chainMaps.get(hash);
```

Comment:
The error handling isn't right here. You need to keep `lock.unlock();` in the finally blocks or otherwise you risk leaking the lock.